backend/lib/spark-api-deal-submitter.js

```diff
@@ -0,0 +1,127 @@
+      console.error('Failed to submit deals:', e)
+    }
+  }
+}
+
+/**
+ * Finds deals that haven't been submitted to spark api.
+ * For deals to be submitted they must
+ * - have not been submitted yet
+ * - were created more than 2 days ago
+ * - have payload cid
+ * - have not yet expired
+ *
+ * @param {PgPool} pgPool
+ * @param {number} batchSize
+ * @returns {AsyncGenerator<Array>}
+ */
+const findUnsubmittedDeals = async function * (pgPool, batchSize) {
+  const client = await pgPool.connect()
+  const cursor = client.query(new Cursor(`
+    WITH two_days_ago AS (
+        SELECT (NOW() - INTERVAL '2 days')::TIMESTAMP AS ts
+    )
+    SELECT
+        miner_id,
+        client_id,
+        piece_cid,
+        piece_size,
+        payload_cid,
+        epoch_to_timestamp(term_start_epoch + term_min) AS expires_at
+    FROM
+        active_deals
+    WHERE
+        submitted_at IS NULL
+        AND payload_cid IS NOT NULL
+        AND activated_at_epoch < timestamp_to_epoch((SELECT ts FROM two_days_ago))
+        AND epoch_to_timestamp(term_start_epoch + term_min) > NOW()`
+  ))
+
+  let rows = await cursor.read(batchSize)
+  while (rows.length > 0) {
+    yield rows
+
+    rows = await cursor.read(batchSize)
+  }
+
+  client.release()
+}
+
+/**
+ * Format unsubmitted deals to format expected by spark api.
+ *
+ * @param {Array} deals
+ * @returns {Array}
+*/
+const formatUnsubmittedDealsForSparkApi = (deals) => {
+  return deals.map(deal => ({
+    minerId: deal.miner_id,
+    clientId: deal.client_id,
+    pieceCid: deal.piece_cid,
+    pieceSize: deal.piece_size.toString(),
+    payloadCid: deal.payload_cid,
+    expiresAt: deal.expires_at
+  }))
+}
+
+/**
+ * Mark deals as submitted.
+ *
+ * @param {Queryable} pgPool
+ * @param {Array} eligibleDeals
+ */
+const markDealsAsSubmitted = async (pgPool, eligibleDeals) => {
+  await pgPool.query(`
+    UPDATE active_deals
+    SET submitted_at = NOW()
+    WHERE miner_id = ANY($1) AND client_id = ANY($2) AND piece_cid = ANY($3) AND piece_size = ANY($4)
+  `, [
+    eligibleDeals.map(deal => deal.miner_id),
+    eligibleDeals.map(deal => deal.client_id),
+    eligibleDeals.map(deal => deal.piece_cid),
+    eligibleDeals.map(deal => deal.piece_size)
+  ])
+}
+
+/**
+ * Submits deals to a spark api.
+ *
+ * @param {string} sparkApiBaseURL
+ * @param {string} dealIngestionAccessToken
+ * @returns {(eligibleDeals: Array) => Promise<void>}
+ */
+export const submitDealsToSparkApi = (sparkApiBaseURL, dealIngestionAccessToken) => async (unsubmittedDeals) => {
+  console.debug(`Submitting ${unsubmittedDeals.length} deals to Spark API`)
+  const response = await fetch(`${sparkApiBaseURL}/eligible-deals-batch`, {
+    method: 'POST',
+    headers: {
+      'Content-Type': 'application/json',
+      Authorization: `Bearer ${dealIngestionAccessToken}`
+    },
+    body: JSON.stringify(unsubmittedDeals)
+  })
+
+  if (!response.ok) {
+    throw new Error(`Failed to submit deals: ${response.statusText}`)
```

Comment:
Please include the status code in the response. I believe the HTTP spec allows the server to return arbitrary status text; it's only the status code number that's standardised.

In my experience, it's difficult to troubleshoot failed requests when we have only the status, because the error details are usually found in the response body.

IMO, it would be best to extract this line into a helper function shared in all places in the backend component.

https://github.com/CheckerNetwork/deal-observer/blob/1c3471def968f3b55302dbc263b1e06c91dec123/backend/lib/rpc-service/service.js#L28